zebra-state/src/memory_state.rs

```diff
@@ -1,3 +1,406 @@
+use std::{
+    cmp::Ordering,
+    collections::BTreeSet,
+    collections::{BTreeMap, HashMap, HashSet},
+    ops::Deref,
+    sync::Arc,
+};
+
+use zebra_chain::{
+    block::{self, Block},
+    primitives::Groth16Proof,
+    sapling, sprout, transaction, transparent,
+    work::difficulty::PartialCumulativeWork,
+    work::difficulty::Work,
+};
+
+use crate::{service::QueuedBlock, BoxError};
+
 pub struct MemoryState {
     // TODO
 }
+
+struct ChainSet {
+    chains: BTreeSet<Chain>,
+
+    queued_blocks: BTreeMap<block::Hash, QueuedBlock>,
+    queued_by_parent: BTreeMap<block::Hash, Vec<block::Hash>>,
+    queued_by_height: BTreeMap<block::Height, Vec<block::Hash>>,
+}
+
+impl ChainSet {
+    pub fn finalize(&mut self) -> Arc<Block> {
+        todo!()
+    }
+
+    pub fn queue(&mut self, block: QueuedBlock) {
+        todo!()
+    }
+
+    fn process_queued(&mut self, new_parent: block::Hash) {
+        todo!()
+    }
+
+    fn commit_block(&mut self, block: QueuedBlock) -> Option<block::Hash> {
+        todo!()
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+struct Chain {
+    blocks: BTreeMap<block::Height, Arc<Block>>,
+    height_by_hash: HashMap<block::Hash, block::Height>,
+    tx_by_hash: HashMap<transaction::Hash, (block::Height, usize)>,
+
+    created_utxos: HashSet<transparent::OutPoint>,
+    spent_utxos: HashSet<transparent::OutPoint>,
+    sprout_anchors: HashSet<sprout::tree::Root>,
+    sapling_anchors: HashSet<sapling::tree::Root>,
+    sprout_nullifiers: HashSet<sprout::Nullifier>,
+    sapling_nullifiers: HashSet<sapling::Nullifier>,
+    partial_cumulative_work: PartialCumulativeWork,
+}
+
+impl Chain {
+    pub fn push(&mut self, block: Arc<Block>) {
+        let block_height = block
+            .coinbase_height()
+            .expect("valid non-finalized blocks have a coinbase height");
+
+        // update cumulative data members
+        self.update_chain_state_with(&block);
+        self.blocks.insert(block_height, block);
+    }
+
+    pub fn pop_root(&mut self) -> Arc<Block> {
+        let block_height = self.lowest_height();
+
+        // remove the lowest height block from self.blocks
+        let block = self
+            .blocks
+            .remove(&block_height)
+            .expect("only called while block is populated");
+
+        // update cumulative data members
+        self.revert_chain_state_with(&block);
+
+        // return the block
+        block
+    }
+
+    fn lowest_height(&self) -> block::Height {
+        self.blocks
+            .keys()
+            .next()
+            .cloned()
+            .expect("only called while block is populated")
+    }
+
+    pub fn fork(&self, new_tip: block::Hash) -> Option<Self> {
+        if !self.height_by_hash.contains_key(&new_tip) {
+            return None;
+        }
+
+        let mut forked = self.clone();
+
+        while forked.non_finalized_tip_hash() != new_tip {
```

Comment:
It might be helpful to call the argument `fork_tip`, to avoid confusion with the new block that's just arrived.

After the fork, we're going to append the new block to the forked chain, and it will become the new tip.
```suggestion
    pub fn fork(&self, fork_tip: block::Hash) -> Option<Self> {
        if !self.height_by_hash.contains_key(&fork_tip) {
            return None;
        }

        let mut forked = self.clone();

        while forked.non_finalized_tip_hash() != fork_tip {
```